charts/cluster-autoscaler/templates/pdb.yaml

```diff
@@ -10,7 +10,10 @@ spec:
   selector:
     matchLabels:
 {{ include "cluster-autoscaler.instance-name" . | indent 6 }}
-{{- if .Values.podDisruptionBudget }}
-  {{ toYaml .Values.podDisruptionBudget | nindent 2 }}
-{{- end }}
+  {{- if .Values.podDisruptionBudget.minAvailable }}
+  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
+  {{- end }}
+  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
```

Comment:
I think having an `{{- else}}` block after `{{- if .Values.podDisruptionBudget.minAvailable }}` here is simpler and equivalent to a separate `{{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}` block.

The way the current values.yaml data structure is implemented is sort of incorrect as a way to represent the podDisruptionBudget interface. See:

- https://kubernetes.io/docs/tasks/run-application/configure-pdb/#specifying-a-poddisruptionbudget

Especially:

> You can specify only one of maxUnavailable and minAvailable in a single PodDisruptionBudget.

I think this PR points towards an improvement in the existing solution, but FYI we're in a non-ideal place here, as we're essentially putting it on the users to understand that they're only allowed to use one or the other of the child properties of the `podDisruptionBudget` values object, and that if they *do* specify both then `minAvailable` will be preferred, and `maxUnavailable` will be ignored. (There isn't a great way to do runtime template error validation in helm AFAIK .)

Reply:
cc @gjtempleton on that last part above

Reply:
> I think this PR points towards an improvement in the existing solution, but FYI we're in a non-ideal place here, as we're essentially putting it on the users to understand that they're only allowed to use one or the other of the child properties of the `podDisruptionBudget` values object, and that if they _do_ specify both then `minAvailable` will be preferred, and `maxUnavailable` will be ignored. (There isn't a great way to do runtime template error validation in helm AFAIK .)

I have generally seen in other helm charts that they usually have the podDisruptionBudget block as empty in values.yaml. By default they set it off, and then allow the users to enter what they want. 

Usually, when users enter both maxUnavailable and minAvailable together in such case, it will add both within the template instead of prioritizing one over above. Applying this generated template will then throw the error by kubernetes itself suggesting that only one of the 2 should be available. I think this is the correct way to go about it. 

The reason I made it prioritize one of the two (maxUnavailable and minAvailable) and did not touch values.yaml was to ensure that it still works the same way for old users. But if that is not the case, then I can change the PR to make it how it should ideally be.

Reply:
Yeah, that all makes sense, and 💯 on making this back-compat for users.

Now that I think about it, maybe this is the most correct template solution:

```
  {{- if .Values.podDisruptionBudget.minAvailable and (not .Values.podDisruptionBudget.maxUnavailable) }}
  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
  {{- end }}
  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: {{ .Values.podDisruptionBudget.maxUnavailable }}
  {{- end }}
```

If we do the above, *if* users declare both configuration flavors, it will be ignored by the template. I suppose this is not strictly back-compat, but the current template will not work in such a scenario anyways (I assume it will be rejected by apiserver webhook for the `PodDisruptionBudget` resource type).

wdyt?

Reply:
Yes, the current template is rejected by the webhook if we declare both flavors. It is also rejected if i just add minAvailable in the override yaml file, because maxUnavailable is included by default in values.yaml.

Reply:
@gjtempleton wdyt?

```
  {{- if .Values.podDisruptionBudget.minAvailable and (not .Values.podDisruptionBudget.maxUnavailable) }}
  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
  {{- end }}
  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: {{ .Values.podDisruptionBudget.maxUnavailable }}
  {{- end }}
```

or picking one or the other and making it highest order precedence, e.g., minAvailable in @ParichayDidwania's original proposal:

```
  {{- if .Values.podDisruptionBudget.minAvailable }}
  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
  {{- end }}
  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: {{ .Values.podDisruptionBudget.maxUnavailable }}
  {{- end }}
```

Reply:
Any update on this @jackfrancis ?

Reply:
After thinking about this I think this is the best, back-compat way forward:

```
  {{- if .Values.podDisruptionBudget.minAvailable and (not .Values.podDisruptionBudget.maxUnavailable) }}
  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
  {{- end }}
  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: {{ .Values.podDisruptionBudget.maxUnavailable }}
  {{- end }}
  {{- if and (not .Values.podDisruptionBudget.maxUnavailable) (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: 1
  {{- end }}
```

And then would need to get rid of the default in values.yaml. I don't think we can statically assign defaults to one or the other based on the mutually exclusive API requirements here; moving the defaults into the template in the above way should solve for that.

Thanks for your patience!

Reply:
@jackfrancis Having defaults in the template is not a good practice in my opinion. This is because, users will always go through the values.yaml and override things from there. Having default should be the job of values.yaml. If someone just turns the pdb on from override.values.yaml, they would automatically end up with maxUnavailable as 1. Helm should only be used for templatizing fields, not providing defaults. 

According to me, we should keep the maxUnavailable in values.yaml for backward compatibility, and just provide a user with a choice to add maxUnavailable or minAvailable from the override.values.yaml. Even if they don't provide anything of the flavours, they would know that maxUnavailable: 1 is applied because that is the default value in values.yaml

What do you think?

Reply:
The problem here is that the values.yaml data interface doesn't support either/or property existence requirements.

> just provide a user with a choice to add maxUnavailable or minAvailable from the override.values.yaml

How do we do the above exactly?

Reply:
@jackfrancis If we have this

` {{- if .Values.podDisruptionBudget.minAvailable }}
  minAvailable: {{ .Values.podDisruptionBudget.minAvailable }}
  {{- end }}
  {{- if and .Values.podDisruptionBudget.maxUnavailable (not .Values.podDisruptionBudget.minAvailable) }}
  maxUnavailable: {{ .Values.podDisruptionBudget.maxUnavailable }}
  {{- end }}`
 
 along with maxUnavailable: 1 in values.yaml. The users will know that they can use override.values.yaml to choose the flavour they want. 
 
 They can simply set in the override, minAvailable: 1 to override the default maxUnavailable. 
 
 In the other case, they can either let it be default and explicitly specify maxUnavailable: 1 in order to get this flavor. 
 
 This way they have a choice. 
 
 Now, if they add both, the ideal scenario is to add both from our end and let the kubernetes API throw an error, but since we need to make it backward compatible, we can't do that.  This is why, we prioritize one over the other. And since, maxUnavailable is present by default, we have to prioritize minAvailable. 
 
 The solution you provided will yield the same result as mine, except the maxUnavailable default value will be visible to users in values.yaml, where the users would expect it to be, and not inside the template itself.

Reply:
@jackfrancis Any update on this?